### ISSUE TITLE
RVM interference with WAR bundled gems in Tomcat 7

### ISSUE THREAD

**Issue by DrPep** [jruby/warbler#263]
Hi,

We encountered some issues with RVM interfering with the gems available to a warbler-packaged WAR file (deployed within Tomcat 7).

Observations:
We generated a WAR with the following warbler configuration:

    Warbler::Config.new do |config|
      config.features = %w(runnable)

      # Application directories to be included in the webapp.
      config.dirs = %w(app config db lib log script vendor doc)
      config.bundle_without = %w(development test assets)
      config.jar_name = 'AppName'
      config.override_gem_home = false
      config.webxml.jruby.min.runtimes = 1
      config.webxml.jruby.max.runtimes = 1
    end

When we set override_gem_home in the warble config to true, the application successfully boots in development and serves pages. However, any rake tasks run against the war file (like db:migrate) fail.

When we set override_gem_home in the warble config to false, rake tasks run successfully, but the application fails to boot.

In both cases, the error given is:

    Bundler::GitError: git@github.com:<user>/<extracted-gem>.git (at master) is not yet checked out. Run `bundle install` first

After trying a number of things, we eventually got both the app and rake tasks to work by setting the RVM gemset in the directory where we were running tomcat to our project's development gemset.

This seems to suggest that RVM is blocking the application from using the gems packaged in the WAR file, in surprising ways. We haven't tracked down the precise mechanism by which this happens, so it's possible that it's an RVM issue rather than something Warbler could deal with.

**Comment by blueboxjesse** [comment]
@jordanvs Did you ever solve this?

**Comment by jordanvs** [comment]
@blueboxjesse Unfortunately not. I remember being stubborn and trying to trial-and-error my way out of it but I never got the rake tasks to work from within the WAR.

My goal was to run rake tasks on the production app server(s) we had deployed from the war file. They were simple tasks and not very intensive so this wasn't a danger to the system resources being utilized by requests. I actually ended up picking up the jruby-quartz gem and integrating that. At least for me, this was quite the effort and some troubleshooting both in development and when it got into production.

My other workaround option would have been to spin up a utility server, use cron/rake, and have that utility server run against the master branch on MRI (i.e. not JRuby/War). I could have built this as an extra step into my build/deploy scripts. It's not pretty and it doesn't make you feel warm and fuzzy, but if you trust your gemset and code to work both in Jruby and MRI, this is at least an option. Useful also to allow a production-pointing `rails c production` with a bit more security potential.

I'm not sure if all that rambling helps you, but I can elaborate on anything if you would find it useful.
